vispy/visuals/instanced_mesh.py

```diff
@@ -0,0 +1,190 @@
+
+from .mesh import MeshVisual
+
+
+_VERTEX_SHADER = """
+uniform bool use_instance_colors;
+
+// these attributes will be defined on an instance basis
+attribute vec3 shift;
+attribute vec4 instance_color;
+attribute vec3 transform_x;
+attribute vec3 transform_y;
+attribute vec3 transform_z;
+
+vec4 color;
+
+varying vec4 v_base_color;
+void main() {
+    if (use_instance_colors) {
+        color = instance_color;
+    } else {
+        color = $base_color;
+    }
+
+    v_base_color = $color_transform(color);
+
+    // transform is generated from column vectors (new basis vectors)
+    // https://en.wikibooks.org/wiki/GLSL_Programming/Vector_and_Matrix_Operations#Constructors
+    mat3 instance_transform = mat3(transform_x, transform_y, transform_z);
+    vec3 pos_rotated = instance_transform * $to_vec4($position).xyz;
+    vec4 pos_shifted = $to_vec4(pos_rotated + shift);
+    gl_Position = $transform(pos_shifted);
+}
+"""
+
+
+class InstancedMeshVisual(MeshVisual):
+    """Instanced Mesh visual.
+
+    Mostly identical to MeshVisual, but additionally takes arrays of
+    of positions and transforms (optionally colors) to create multiple
+    instances of the mesh.
+
+    Parameters
+    ----------
+    vertices : array-like | None
+        The vertices.
+    faces : array-like | None
+        The faces.
+    vertex_colors : array-like | None
+        Colors to use for each vertex.
+    face_colors : array-like | None
+        Colors to use for each face.
+    color : instance of Color or ColorArray
+        The color to use.
+    vertex_values : array-like | None
+        The values to use for each vertex (for colormapping).
+    meshdata : instance of MeshData | None
+        The meshdata.
+    shading : str | None
+        Shading to use. This uses the
+        :class:`~vispy.visuals.filters.mesh.ShadingFilter`
+        filter introduced in VisPy 0.7. This class provides additional
+        features that are available when the filter is attached manually.
+        See 'examples/basics/scene/mesh_shading.py' for an example.
+    mode : str
+        The drawing mode.
+    **kwargs : dict
+        Keyword arguments to pass to `Visual`.
+
+    Notes
+    -----
+    Additional functionality is available through filters. Mesh-specific
+    filters can be found in the :mod:`vispy.visuals.filters.mesh` module.
+
+    This class emits a `data_updated` event when the mesh data is updated. This
+    is used for example by filters for synchronization.
+
+    Examples
+    --------
+    Create a primitive shape from a helper function:
+
+    >>> from vispy.geometry import create_sphere
+    >>> meshdata = create_sphere()
+    >>> mesh = MeshVisual(meshdata=meshdata)
+
+    Create a custom shape:
+
+    >>> # A rectangle made out of two triangles.
+    >>> vertices = [(0, 0, 0), (1, 0, 1), (1, 1, 1), (0, 1, 0)]
+    >>> faces = [(0, 1, 2), (0, 2, 3)]
+    >>> mesh = MeshVisual(vertices=vertices, faces=faces)
+    """
```

Comment:
Can you update this docstring with the new keyword arguments and any usage examples that you think make sense (remove the existing ones that were copied from MeshVisual).

Reply:
Oh a link to a site describing instanced rendering would be really nice.

Reply:
Yeah maybe just a nice high-level human-readable description of what instanced rendering is would be nice. I'm basically imagining someone coming to VisPy and knowing very little about OpenGL, buffers, textures, or even only a little bit about what makes up a mesh and they question the difference between these two Visuals and why they should choose one over the other.